[#55581] popover for user information on hover

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/55581

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

User hover cards when hovering over the avatar. I tried also showing the hover card when hovering over the name, but it felt wonky and overly verbose. EDIT: as discussed in the frontend daily, the wonkyness is no longer present with an increased delay before showing a hover card. So showing it for names is an option now. We decided that I will create an open point for this topic so that we can decide how to proceed with this. In case we want to add this, I would supply another PR for that, as it will be much smaller in scope.

Please take extra care when reviewing this. I have tried to check every place where a user principal is rendered, but might have missed some due to the sheer number of them.

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

![image](https://github.com/user-attachments/assets/b2e6aa0a-c72b-4129-af83-f8396d64edf0)


https://github.com/user-attachments/assets/73f3f4a2-3d46-4e65-be98-930bc724b315

_Note_: the video is not completely accurate. The arrow pointing downwards towards the avatar has been removed because it's not possible to position it reliably.

# What approach did you choose and why?

Reused the HoverCard component from work packages and made some adjustments.

Some changes made to the trigger service (Typescript) also affect the work packages. This includes a longer delay before a hover card appears, as well as a slightly longer delay before it closes. Bugs happening when you quickly over multiple different hover cards have been fixed.

The `AvatarComponent` as well as the `op-principal-component` now support an option for enabling hover cards. By default, the popover is **enabled**. I had to _disable_ the popover in some places where it caused hard-to-fix bugs, such as: it is not possible to render another spot modal within an existing spot modal.

These places include:

* the global search result dropdown
* the time tracking modal
* the work package share modal

# Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#55581] enforce hover card in project member list for now

EinLama committed:
[#55581] unwrap quotes for angular components

EinLama committed:
[#55581] return early

EinLama committed:
[#55581] Users::HoverCardComponent

EinLama committed:
[#55581] prefer flex layout

It is closer to what we want to achieve.

EinLama committed:
WIP

EinLama committed:
[#55581] offer hover card trigger switch in avatar helper

EinLama committed:
[#55581] first draft for triangle pointing at hover source

EinLama committed:
[#55581] construct project membership string, link to profile

This is only a draft since the links are still missing.
But it gives a good first impression of what's to come.

EinLama committed:
avatar details

We might have to patch a bigger possible size here later,
the default size is currently the biggest and it looks a
bit lost in the hover card.

EinLama committed:
[#55581] show email if allowed to do so

EinLama committed:
[#55581] link project names to projects

EinLama committed:
[#55581] attach hover card attributes to avatar image

Previously, it was attached to the wrapper, which led to weird behavior.
With the new approach, it looks more natural while fitting the design
better.

Sneaked in some improvements such as using the hover card path helper.

EinLama committed:
[#55581] copyright notices

EinLama committed:
[#55581] clear timeout for closing hover card

This allows to hover over cards in quick succession without
them interfering with each other.

EinLama committed:
return as soon as possible

EinLama committed:
[#55581] set position & close delay of card by data-attr

EinLama committed:
[#55581] do not show the same modal twice

Doing so causes flickering. We don't want that as it looks weird.

EinLama committed:
[#55581] link to groups instead of projects

EinLama committed:
[#55581] link the end of the sentence to users#show

EinLama committed:
[#55581] allow hovercards in auto completer and share dialog

EinLama committed:
[#55581] only show hovercard if set active

EinLama committed:
[#55581] do not attempt to show hover cards for groups

EinLama committed:
[#55581] ensure hover card is rendered in front of modals

EinLama committed:
[#55581] show hover card in member list

EinLama committed:
WIP

EinLama committed:
[#55581] make portal target customizable

EinLama committed:
default should be default

EinLama committed:
[#55581] provide hover card in news module

EinLama committed:
[#55581] hover card on boards

EinLama committed:
[#55581] hover card in wp activities

EinLama committed:
[#55581] hover card in meetings

EinLama committed:
[#55581] hover card on project overview page (members widget)

EinLama committed:
[#55581] fix z-index on top of auto completers

EinLama committed:
[#55581] small polishing fixes

EinLama committed:
[#55581] do not set top alignment by default

Since rebasing on latest dev, this causes issues with the
mousedetection. Revert to default behaviour by default for now.

HDinger commented on app/components/users/hover_card_component.html.erb:
```diff
@@ -0,0 +1,77 @@
+
+            if show_email?
+              f.with_column do
+                render(Primer::Beta::Text.new(color: :muted)) { @user.mail } # TODO: only with permission
```
I guess the `TODO` can be removed?

EinLama replied:
Good find. Removed!

EinLama committed:
[#55581] add necessary translation for hover card permission

EinLama committed:
[#55581] remove redundant comments, provide 404 locale

EinLama committed:
frozen_string_literal

EinLama commented on config/initializers/permissions.rb:
```diff
@@ -71,6 +71,12 @@
                      require: :loggedin,
                      contract_actions: { users: %i[read update] }
 
+      map.permission :view_user_hover_card,
+                     {
+                       "users/hover_card": %i[show]
+                     },
+                     permissible_on: :global
+
```
Really not sure about this one. We *need* to provide some sort of permission when we want to render a component for the hover cards. But which one should it be?

EinLama replied:
Removed in favor of `no_authorization_required!`

EinLama committed:
[#55581] explain missing urls in some cases

EinLama committed:
[#55581] remove unneeded permission

EinLama committed:
mv project group

EinLama committed:
[#55581] do not show hover card for placeholder users

HDinger reviewed (changes requested):
Hi @EinLama 

as discussed, the amount of cards feels rather annoying. I'd like the challenge the requirement that the card should really be shown all the time. Alternatively, the delay must be larger. But that has to be discussed with the product team.

Until then, here are the first findings from clicking through:

- [x] Hovering a placeholder user leads to an error
- [x] As a non-admin user, there is an error as well due to a missing permission
- [x] Opening the share modal on a WP creates a black overlay
- [x] When the LogTime modal on a WP is opened, and the user dropdown is opened: Hovering the avatar, will close the logTime modal
- [x] The emailAdress is supposed to smaller (12px)
- [x] When having a very long name, it overflows the card without truncation or wrapping:
<img width="400" alt="Bildschirmfoto 2024-11-22 um 14 11 40" src="https://github.com/user-attachments/assets/6d121900-3f0d-4e58-bc52-8f6308e52781">

- [x] When the hovercard is opened at the very bottom of the page, it closes immideatly

https://github.com/user-attachments/assets/3176db4e-369b-44bb-ac6b-1b5e0769324f


For the record, there are some places where the card is not shown yet. But as said above, I'd wait until the decision of the product team
* Team planner module
* Admin/user table
* Project overview / project attribute of type user
* Project activity page
* Users page (`/users/:id`) 
* My Page / News widget
* Multi-User custom field

EinLama committed:
[#55581] reduce ABC in avatar helper

EinLama committed:
[#55581] spec for user hover cards on member page

EinLama committed:
[#55581] reduce email font size

EinLama committed:
[#55581] do not let very long names/mails overflow

HDinger commented on app/components/users/avatar_component.rb:
```diff
@@ -32,14 +32,16 @@ class AvatarComponent < ApplicationComponent
 
-    def initialize(user:, show_name: true, link: true, size: "default", classes: "", title: nil, name_classes: "")
+    def initialize(user:, show_name: true, link: true, size: "default", classes: "", title: nil, name_classes: "",
+                   hover_card: { active: false, target: :default })
```
Given the current requirement, that the card should be shown on all avatars, wouldn't it be better to set `active` to true per default?

EinLama replied:
Yes, you are right; I will flip the default bit!

HDinger commented on app/components/users/hover_card_component.html.erb:
```diff
@@ -0,0 +1,76 @@
+        end
+
+        flex.with_row do
+          # TODO: could also use helpers.allowed_management_user_profile_path here...
```
Open Todo

HDinger commented on app/components/users/hover_card_component.html.erb:
```diff
@@ -0,0 +1,76 @@
+        flex.with_row do
+          # TODO: could also use helpers.allowed_management_user_profile_path here...
+          render(Primer::Beta::Button.new(tag: :a, href: user_path(@user))) do
+            "Open profile"
```
Please localise that string

HDinger commented on app/components/users/hover_card_component.rb:
```diff
@@ -0,0 +1,95 @@
+  # or
+  # "Member of group1, group2 and 3 more"
+  # The latter string is cut off since the complete list of group names would exceed the allowed `max_length`.
+  def group_membership_summary(max_length = 40)
```
Where is the value 40 coming from?

EinLama replied:
A true magic number 🔮🧙🏻‍♂️ found by, uh... guessing and trying it out! 🪄 

Seemed sensible when used with the English translation and should look okay with most other languages. The caveat is that only the group names are considered when deciding where to shorten the list. To make it perfect, we'd have to load the translation first, factor it's length into the calculation and THEN decide where to cut off the group list. I figured the value `40` is good enough in most cases and strikes a balance between pleasant looks and code complexity :)

HDinger commented on app/components/users/hover_card_component.sass:
```diff
@@ -0,0 +1,23 @@
+    gap: 0.5rem
+
+    .op-user-hover-card--name, .op-user-hover-card--email
+      overflow: hidden
+      text-overflow: ellipsis
+
```
We have a mixin for this called `text-shortener`

EinLama replied:
Oh, nice! That's convenient. Added.

HDinger commented on spec/features/members/membership_spec.rb:
```diff
@@ -138,6 +138,22 @@
+        button = card.find(".Button", text: "Open profile")
+        expect(button["href"]).to eq(user_url(peter))
+      end
+    end
```
Please use `test_selectors` for all of this. You can also use the helpers `within_test_selector` and `find_test_selector`

HDinger commented on frontend/src/app/shared/components/principal/principal.component.ts:
```diff
@@ -76,6 +74,14 @@ export class OpPrincipalComponent implements OnInit {
 
+  @Input() avatarClasses? = '';
+
+  @Input() hoverCard= false;
```
Same as in the rails code: Why is the default `false` if it should be shown all the time?

HDinger commented on frontend/src/app/shared/components/fields/display/field-types/user-display-field.module.ts:
```diff
@@ -48,7 +48,7 @@ export class UserDisplayField extends DisplayField {
         this.attribute,
         { hide: false, link: false },
-        { hide: false, size: 'medium' },
+        { hide: false, size: 'medium', hoverCard: { url: this.attribute.url } },
```
ESLint is somehow complaining correctly here.`this.attribute` can be anything, so we do not know for sure whether a `url` is defined. Ideally this case would at least be catched before.

EinLama replied:
There were other issues with this class, unrelated to this PR. I have made an attempt at fixing them. The compiler is happy, so far is the runtime. Will observe if this causes problems somewhere in the application.

HDinger commented on frontend/src/app/core/setup/globals/global-listeners/hover-card-trigger.service.ts:
```diff
@@ -29,12 +29,20 @@
   private mouseInModal = false;
+  private closeTimeout:number|null = null;
+  private closeDelayInMs:number = 100;
+  private modalAlignment:string|null = null;
```
You probably do not want to allow all strings here. Further down, you parse it to type `Placement`, so why don't you require `Placement|null` from the beginning and do a proper check when you parse it from the data attributes? Otherwise, nothing prevents a developer from passing somehting wrong and thus creating a potential bug

HDinger commented on frontend/src/app/shared/components/principal/principal-renderer.service.ts:
```diff
@@ -6,12 +6,21 @@ import idFromLink from 'core-app/features/hal/helpers/id-from-link';
+  url?:string;
+  modalTarget?:PortalOutletTarget;
+  closeDelay?:number;
+  alignment?:string;
```
Same as before: Alginment should be of type `Placement` and not any string

HDinger commented on frontend/src/app/shared/components/principal/principal-renderer.service.ts:
```diff
@@ -6,12 +6,21 @@ import idFromLink from 'core-app/features/hal/helpers/id-from-link';
+export interface HoverCardOptions {
+  url?:string;
+  modalTarget?:PortalOutletTarget;
+  closeDelay?:number;
```
Why does the the frontend has the option for custom delays and alignments? It feels like it is overcomplicating things.

EinLama replied:
I think I can drop the alignment. It seemed useful in an earlier iteration where I wanted to stay closer to the Figma design and center the hover card over the avatar. However, it doesn't look as good as I figured it should and later on I decided to treat WP and user hover cards similar.

The closeDelay has a similar reasoning. I wanted the user hover card to linger around longer since it's easy to accidentally leave the avatar on many pages. At the same time, I did not want to change WP hover card behavior: hence a new option.

I could just increase the closeDelay for both WP and user hover cards and remove this complexity. That would probably be best 🤔

EinLama replied:
We discussed this and decided to treat WP and user hover cards the same. I have removed both options to reduce the complexity 👍🏻

HDinger commented on app/components/users/hover_card_component.sass:
```diff
@@ -0,0 +1,23 @@
+// When displaying the hover card on top of an existing modal, we must
+// set the z-index so that the hover card stays in the front.
+.op-user-share-modal-overlay
+  position: absolute
+  z-index: 9600
+
+// Correct the z-index of the regular hover card container so that it is above user auto completers
+.spot-modal-overlay:has(.op-user-hover-card)
+  z-index: 9600
```
Despite the comments I don't really get why this is needed (or rather why other implementations with a custom overlay don't need it). Maybe we can shortly talk about this.

HDinger reviewed (commented):
It feels like you could save a lot of code by turning the default for the hoverCards to `true`. Is there a reason it is set to `false`?

EinLama committed:
[#55581] wait for one second before displaying a hover card

This makes the hover card feel less annoying when letting the mouse
pointer wander across a page.

EinLama committed:
[#55581] remove superfluous styling

EinLama committed:
[#55581] link to edit profile, localize button string

EinLama committed:
[#55581] use text-shortener mixin

EinLama committed:
[#55581] remove alignment and close delay options

We will treat all hover cards the same to reduce the complexity.

EinLama committed:
[#55581] make user hover card opt-out

By default, it will now be shown. Since we want to have it almost
anywhere on the page, it makes sense that `true` is the default setting.

EinLama committed:
fix indentation

EinLama committed:
[#55581] allow disabling the hover card

EinLama committed:
[#55581] disable hover card on hover card

It's hover cards all the way down.

EinLama committed:
[#55581] only show visible groups within the hover card

To align with the user profile. We don't want to show hidden groups here
and expose secret information.

EinLama committed:
[#55581] hover cards only in project share, not wp share

They conflict with each other, so we disable them.

EinLama committed:
[#55581] fix bug where hover cards are sometimes directly dismissed

This happened when the hover card rendered on top (e.g. when you
scrolled down a bit and there was no space below the trigger). The hover
card will now also consider the trigger when deciding whether the mouse
left the card or not.

EinLama committed:
[#55581] do not try to show a modal twice at the same time

There is only one possible slot to inject a modal. If that slot is
taken, abort!

EinLama committed:
[#55581] use test selectors in spec

EinLama committed:
[#55581] specs for Users::HoverCardComponent

EinLama committed:
[#55581] fix avatar_helper specs

EinLama commented:
Thank you, @HDinger. I have fixed most of the issues and findings. I found another place that triggers erratic behavior:

- [x] global search closes when hovering over an avatar (e.g. work package authors)

EinLama committed:
[#55581] do not show the hover card within the log time dialog

EinLama committed:
[#55581] ensure the URL is a string

EinLama committed:
[#55581] disable hover cards in global search

EinLama committed:
[#55581] lookbook proposal

EinLama committed:
[#55581] add exemplary user hover card to lookbook

EinLama committed:
[#55581] reduce filesize of wp hover card image by 45%

compressed via ImageOptim

EinLama committed:
[#55581] user generic error message for missing users

EinLama committed:
[#55581] do not try to access non-existent field

EinLama committed:
[#55581] make hover card work with avatar fallbacks

This was a hard one to debug. When you hover over an avatar as soon as
the page loads, you will hover over the fallback first. Within the
timeout, the actual avatar image loads and replaces the fallback. After
that, the timeout of the hover trigger runs out and displays the hover
cards, but does no longer have an element to position relative to. So,
we check for the existence of the hovered element before showing a hover
card to fix this.

EinLama committed:
[#55581] animate hover card slightly to make it feel less awkward

EinLama committed:
[#55581] deal with multiple hover triggers next to each other

If you quickly move from one hover target to the next, we know reset the
hover card component into the correct state so that you can open the
next hover card. Just keep hovering.

EinLama committed:
[#55581] prevent background layer from darkening in multi nested dialogs

EinLama committed:
Merge branch 'dev' into feature/55581-popover-for-user-information-on-hover

oliverguenther commented on app/components/users/hover_card_component.sass:
```diff
@@ -0,0 +1,24 @@
+  gap: 1rem
+  overflow: hidden
+
+  .op-user-hover-card--info
```
this could be just `&--info`. This ensures that you're using the BEM style

oliverguenther commented on app/components/users/hover_card_component.sass:
```diff
@@ -0,0 +1,24 @@
+  gap: 1rem
+  overflow: hidden
+
+  .op-user-hover-card--info
```
```suggestion
  &--info
```

this could be just &--info. This ensures that you're using the BEM style

EinLama replied:
Very nice, thanks! The previous class definitions were very cumbersome. Much better like this.

oliverguenther commented on app/components/users/hover_card_component.sass:
```diff
@@ -0,0 +1,24 @@
+  .op-user-hover-card--info
+    gap: 0.5rem
+
+    .op-user-hover-card--name, .op-user-hover-card--email
```
```suggestion
    &--name, &--email
```

same as above

oliverguenther commented on app/components/users/hover_card_component.sass:
```diff
@@ -0,0 +1,24 @@
+    .op-user-hover-card--name, .op-user-hover-card--email
+      @include text-shortener()
+
+  .op-user-hover-card--group-list
```
```suggestion
  &--group-list
```

oliverguenther commented on frontend/src/app/shared/components/principal/principal-renderer.service.ts:
```diff
@@ -170,6 +191,11 @@ export class PrincipalRendererService {
 
+  private userHoverCardUrl(principal:PrincipalLike|IPrincipal):string|null {
+    const id = principal.id || idFromLink(hrefFromPrincipal(principal));
+    return id ? `/users/${id}/hover_card` : null;
```
This will fail in case of relative URL root installations. Please use the `PathHelperService` to generate static links, they are correctly using the appBase/releative URL root path

EinLama replied:
Good point, thanks! Changed ✅

oliverguenther commented on frontend/src/app/core/setup/globals/global-listeners/hover-card-trigger.service.ts:
```diff
@@ -49,31 +60,48 @@ export class HoverCardTriggerService {
+      if (!turboFrameUrl) { return; }
+
+      // When set in an angular component, the url attribute might be wrapped in additional quotes. Strip them.
+      const cleanedTurboFrameUrl = turboFrameUrl.replace(/^"(.*)"$/, '$1');
```
The reason this happens is because you use the inputs that are used for components in the AvatarHelper, which is not too useful for services.

These often get read in this helper function:
https://github.com/opf/openproject/blob/c757e6f60e68c995f6c17db52fa379fb5b62aeb7/frontend/src/app/shared/components/dataset-inputs.ts#L80

and in the end, it's JSON. Of course replacing it might be fine, but `JSON.parse` would also work. What I don't understand: You get the `hoverCardUrl` in the principal component, so it's parsed correctly there. Shouldn't this be used to render the avatar? In case of work packages, it seems a plain data-attribute is already used.

Are these lines still necessary?

EinLama replied:
Yes, we actually don't need these lines anymore. This is a left over from a time earlier during development when I applied the trigger to the principal-tag directly. This was done to support having a popover on both the avatar and the name (and possibly the gap in-between them). Back then, these data-attributes were read from the principal angular tag by the trigger service - breaking it.

Since the trigger is now attached to the avatar image tag, encountering a broken data-url is no longer an issue and this code can be removed 🧹

oliverguenther reviewed (changes requested):
Behavior-wise, LGTM. I have some minor remarks

EinLama committed:
[#55581] use BEM style class definitions

EinLama committed:
[#55581] use PathHelperService to construct hover card url

EinLama committed:
[#55581] remove unneede clean up of angular double quotes

This was a leftover that can now be removed.
